man/vwait.8
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.41.2.
.TH VWAIT "8" "May 2013" "vwait  -- waits for a context to finish" "System Administration"
.SH NAME
vwait \- waits for a context to finish
.SH SYNOPSIS
.B vwait
[\fI--timeout <timeout>\fR] [\fI--terminate\fR] [\fI--status-fd <fd>\fR] [\fI--\fR] \fI<xid>\fR
.SH DESCRIPTION
Please report bugs to vserver@list.linux\-vserver.org
.PP
This program is part of util\-vserver
.SH COPYRIGHT
Copyright \(co 2005 Enrico Scholz
This program is free software; you may redistribute it under the terms of
the GNU General Public License.  This program has absolutely no warranty.


NEWS
version 0.30.215
================

	- added support for using libnss instead of beecrypt for
	vhashify's hash calculation.  libnss has bad SHA1 performance on
	i386, but is a) maintained and b) gives better performance with
	all other hashes and on x86_64.

	- added tunctl, a small program to setup tun/tap interfaces, with
	support in the configuration.

	- vattribute and nattribute now support a --get operation, which
	parses the flags and capabilities into a human-readable form.

	- vspace supports all the spaces currently available.

	- vmemctrl controls the OOM bias of context.

	- dynamic contexts are unbroken.

	- hopefully all rpm-fake races have been fixed.

	- flock has been replaced by lockf, which works on NFS.

	- the clone build method now supports exclude lists.

	- vmount handles mounting into a guest.

	- various fixes for the Gentoo-scripts, thanks to Benedikt Böhm.

	- support for yum's persistdir has been added.

	- per-guest device maps are now supported.

	- vserver ... build now sets the barrier on the parent directory.

	- fsck is now supported for the guest's fstab.

	- symlinks are now handled on all levels in namespace cleanup.

	- the debootstrap build method now supports debootstrap 1.0.8.

	- a documentation update, thanks to Micah Anderson.

	- use numeric uids/gids for the template build method.

	- add support for the alpine and f8 distributions.
	
version 0.30.214
================
	- patches for yum-3.2.

	- various fixes to allow for *SuSE guests.

	- added Fedora 7.

	- replace file with h2ext for build -m template.

	- add exclude list support for vclone.

	- include debootstrap scripts for a lot of distributions.

	- fixes to allow the Gentoo initscripts to be used with baselayout 2.

	- remove disk limit caches on vserver ... delete.

	- added support for the new APIs in 2.2 and 2.3, in particular the
	netv2 API, allowing native IPv6 support.

	- added support for a separate filesystem tag.

version 0.30.213
================
	- added support for the device mapping feature present in development
	kernels.

	- further improved Gentoo support, again thanks to Benedikt Böhm.

	- support persistent contexts.

	- add another initscript that creates directories, configures vshelper
	as well as kills all contexts.

	- new vserver-stat implementation, which uses the accounting APIs
	available in Linux-VServer 2.2.

	- added support for setting sysctl values in the guest configuration.

	- added a clone build method, which simply links already unified files.

	- implement dynamic contexts in userspace, generating a static one
	during vserver ... build.

	- fix the endless loop in vlogin.

	- fix Ubuntu builds.

	- improved support for Debian Etch, thanks to Benedikt Böhm.

	- added CentOS 5.

	- use the state change helper to try to cleanup after us if the guest
	exited on its own.

version 0.30.212
================
	- support for the Linux-VServer 2.1 API has been implemented.

	- the scheduler configuration has been split up into multiple files.

	- namespace cleanup is enabled by default. This can be disabled by
	touch ${sysconfdir}/vservers/.defaults/nonamespace-cleanup

	- all unset ulimits are now reset to infinity, so limits are no longer
	inherited from the host.

	- much improved Gentoo support, thanks to Benedikt Böhm.

version 0.30.211
================

	- novlandev is now the default. To reenable the old behaviour of
	creating VLAN interfaces where used, touch
	${sysconfdir}/vservers/.defaults/interfaces/vlandev.

	- vshelper now works correctly with the kernel legacy support turned
	off.

	- vlogin provides a terminal proxy which allocates a new psuedo-tty
	on the inside when you enter a guest.

	- and lots of other features/fixes, see ChangeLog.


version 0.30.210
================

	- LEGACY BREAKAGE: by default, only recent kernel APIs will be
	supported so you have to make sure to use at least the vs2.0.1
	kernel patch. With older kernels, you have to call 'configure'
	with '--enable-apis=NOLEGACY'. Some programs will not work anymore
	with the default settings; these ones are:

	     + capchroot
	     * chcontext-context
	     * reducecap

	- dlimit support was integrated into the standard vserver
	operations. Thanks to Daniel Hokka Zakrisson for his patch.

	- recent kernel API is now used for setting up the network
	contexts. So it should be now possible to use a kernel compiled
	without CONFIG_VSERVER_LEGACYNET. Thanks to Daniel Hokka Zakrisson
	for his patch.  .


version 0.30.209
================

	- large revamping...
	
	
version 0.23.6
==============

	- fixed '--level' option on 'vserver ... chkconfig'


src/vwait.c
// $Id$    --*- c -*--

// Copyright (C) 2005 Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>
//  
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; version 2 of the License.
//  
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//  
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include "lib/vserver.h"
#include "lib/internal.h"
#include "util.h"

#include <getopt.h>
#include <signal.h>
#include <time.h>
#include <errno.h>
#include <libgen.h>
#include <assert.h>

#define ENSC_WRAPPERS_PREFIX	"vwait: "
#define ENSC_WRAPPERS_STDLIB	1
#define ENSC_WRAPPERS_UNISTD	1
#define ENSC_WRAPPERS_VSERVER	1
#include <wrappers.h>

#define CMD_HELP	0x8000
#define CMD_VERSION	0x8001

#define CMD_TIMEOUT	0x4000
#define CMD_TERMINATE	0x4001
#define CMD_STATUS_FD	0x4002

static struct option const
CMDLINE_OPTIONS[] = {
  { "help",       no_argument,        0, CMD_HELP },
  { "version",    no_argument,        0, CMD_VERSION },
  { "timeout",    required_argument,  0, CMD_TIMEOUT },
  { "terminate",  no_argument,        0, CMD_TERMINATE },
  { "status-fd",  required_argument,  0, CMD_STATUS_FD },
  { 0,0,0,0 }
};

int			wrapper_exit_code = 1;
static sig_atomic_t	aborted = 0;

struct StatusType {
    enum {stERROR, stFINISHED, stKILLED,
	  stTIMEOUT}				status;
    int						rc;
};

struct Arguments
{
    xid_t		xid;
    int			timeout;
    int			status_fd;
    bool		do_terminate;
};

static void
showHelp(char const *cmd)
{
  VSERVER_DECLARE_CMD(cmd);

  WRITE_MSG(1, "Usage:  ");
  WRITE_STR(1, cmd);
  WRITE_STR(1,
	    " [--timeout <timeout>] [--terminate] [--status-fd <fd>] [--] <xid>\n"
	    "\n"
	    "Please report bugs to " PACKAGE_BUGREPORT "\n");
  exit(0);
}

static void
showVersion()
{
  WRITE_MSG(1,
	    "vwait " VERSION " -- waits for a context to finish\n"
	    "This program is part of " PACKAGE_STRING "\n\n"
	    "Copyright (C) 2005 Enrico Scholz\n"
	    VERSION_COPYRIGHT_DISCLAIMER);
  exit(0);
}

static void
handler(int UNUSED num)
{
  aborted = 1;
}

static struct StatusType
doit(struct Arguments const *args)
{
  time_t			end_time = 0, now = 0;
  struct StatusType		res;
  
  if (args->timeout>0) {
    end_time = time(0) + args->timeout;
    siginterrupt(SIGALRM, 1);
    signal(SIGALRM, handler);
    alarm(args->timeout);
  }

  for (;;) {
    res.rc = vc_wait_exit(args->xid);
    
    if      (res.rc==-1 && errno!=EAGAIN && errno!=EINTR) {
	// the error-case
      res.rc     = errno;
      res.status = stERROR;
      perror(ENSC_WRAPPERS_PREFIX "vc_wait_exit()");
    }
    else if (res.rc==-1 && args->timeout>0 && (now=time(0))>=end_time) {
	// an EINTR or EAGAIN signal was delivered, a timeout was set and
	// reached
      if (!args->do_terminate)
	res.status = stTIMEOUT;
      else {
	vc_ctx_kill(args->xid, 1, 9);
	vc_ctx_kill(args->xid, 0, 9);
	res.status = stKILLED;
      }
    }
    else if (res.rc==-1) {
	// an EINTR or EAGAIN signal was delivered but the timeout not set or
	// not reached yet

	// we are here, when args->timeout==0 or 'now' was initialized (and
	// compared with 'end_time'). So, 'now' can be used below.
      assert(args->timeout<=0 || (now < end_time));

      if (args->timeout>0)	// (re)set the alarm-clock
	alarm(end_time-now);

      continue;
    }
    else
	// vc_wait_exit(2) finished successfully
      res.status = stFINISHED;

    break;
  }

  alarm(0);
  return res;
}

static void
writeStatus(int fd, char const *str, int const *rc, int exit_code)
{
  if (fd==-1) exit(exit_code);

  WRITE_STR(fd, str);
  if (rc) {
    char		buf[sizeof(*rc)*3 + 2];
    size_t		len = utilvserver_fmt_long(buf, *rc);
    WRITE_MSG(fd, " ");
    Vwrite   (fd, buf, len);
  }
  WRITE_MSG(fd, "\n");
  
  exit(exit_code);
}

int main(int argc, char *argv[])
{
  struct StatusType	res;
  struct Arguments	args = {
    .xid          = VC_NOCTX,
    .timeout      = -1,
    .status_fd    = -1,
    .do_terminate = false,
  };

  while (1) {
    int		c = getopt_long(argc, argv, "c:", CMDLINE_OPTIONS, 0);
    if (c==-1) break;

    switch (c) {
      case CMD_HELP		:  showHelp(argv[0]);
      case CMD_VERSION		:  showVersion();
      case CMD_TERMINATE	:  args.do_terminate = true;         break;
      case CMD_TIMEOUT		:  args.timeout      = atoi(optarg); break;
      case CMD_STATUS_FD	:  args.status_fd    = atoi(optarg); break;
      default		:
	WRITE_MSG(2, "Try '");
	WRITE_STR(2, argv[0]);
	WRITE_MSG(2, " --help' for more information.\n");
	return EXIT_FAILURE;
	break;
    }
  }

  if (optind+1 > argc) {
    WRITE_MSG(2, ENSC_WRAPPERS_PREFIX "no context specified; try '--help' for more information\n");
    exit(wrapper_exit_code);
  }

  if (optind+1 < argc) {
    WRITE_MSG(2, ENSC_WRAPPERS_PREFIX "can not wait for more than one context; try '--help' for more information\n");
    exit(wrapper_exit_code);
  }
  
  args.xid = Evc_xidopt2xid(argv[optind], true);

  if (args.xid==VC_NOCTX) {
    WRITE_MSG(2, ENSC_WRAPPERS_PREFIX "invalid context specified; try '--help' for more information\n");
    exit(wrapper_exit_code);
  }

  res = doit(&args);

  switch (res.status) {
    case stERROR	:  writeStatus(args.status_fd, "ERROR",    &res.rc, 127);
    case stFINISHED	:  writeStatus(args.status_fd, "FINISHED", &res.rc,   0);
    case stKILLED	:  writeStatus(args.status_fd, "KILLED",         0,   1);
    case stTIMEOUT	:  writeStatus(args.status_fd, "TIMEOUT",        0,   2);
    default		:  writeStatus(args.status_fd, "???",      &res.rc, 126);
  }
}
